File: Problems/Algorithms/799. Champagne Tower/ChampagneTower.java
class Solution {
    public double champagneTower(int poured, int query_row, int query_glass) {
        double[][] res = new double[102][102];
        res[0][0] = (double) poured;

        for (int row = 0; row <= query_row; row++) {
            for (int col = 0; col <= row; col++) {
                double curr = (res[row][col] - 1.0) / 2.0;
                if (curr > 0) {
                    res[row+1][col] += curr;
                    res[row+1][col+1] += curr;
                }
            }
        }

        return Math.min(1, res[query_row][query_glass]);
    }
}

File: Problems/Algorithms/799. Champagne Tower/ChampagneTower.kt
class Solution {
    fun champagneTower(poured: Int, query_row: Int, query_glass: Int): Double {
        var curr = DoubleArray(100) { 0.0 }
        curr[0] = poured.toDouble()
        
        for (i in 0..query_row-1) {
            val next = DoubleArray(100) { 0.0 }
            for (j in 0..99) {
                if (curr[j] > 1.0) {
                    val rest = curr[j] - 1.0
                    curr[j] = 1.0
                    next[j] += rest / 2.0
                    next[j+1] += rest / 2.0
                }
            }
            
            curr = next
        }
        
        return minOf(1.0, curr[query_glass])
    }
}

